config.org
#+TITLE: Emacs Configuration

* General Settings

** Set package archives

   #+BEGIN_SRC emacs-lisp

     (setq package-archives
	   '(("elpy" . "http://jorgenschaefer.github.io/packages/")
	     ("melpa" . "https://melpa.org/packages/")
	     ("gnu" . "http://elpa.gnu.org/packages/")
	     ("melpa-stable" . "https://stable.melpa.org/packages/")))

   #+END_SRC 
   
** Set default tab width

   #+BEGIN_SRC emacs-lisp

     (setq default-tab-width 2)

   #+END_SRC 

** Don't show startup messages

   #+BEGIN_SRC emacs-lisp

     (setq inhibit-startup-screen t)

   #+END_SRC 

** Don't show toolbar

   #+BEGIN_SRC emacs-lisp

     (tool-bar-mode -1)

   #+END_SRC 

** Don't show menu

   #+BEGIN_SRC emacs-lisp

     (menu-bar-mode -1)

   #+END_SRC 

** Don't show scrollbar

   #+BEGIN_SRC emacs-lisp

     (toggle-scroll-bar -1)

   #+END_SRC 

** Disable bell

   #+BEGIN_SRC emacs-lisp

     (setq visible-bell nil)

   #+END_SRC 

** Show matching parenthesis 

   #+BEGIN_SRC emacs-lisp

     (show-paren-mode 1)

   #+END_SRC 

** Enable global line highlight mode

   #+BEGIN_SRC emacs-lisp

     (global-hl-line-mode +1)

   #+END_SRC 

** Enable global line number mode

   #+BEGIN_SRC emacs-lisp

     (add-hook 'prog-mode-hook 'display-line-numbers-mode)

   #+END_SRC 

** Create command to reload configuration 

   #+BEGIN_SRC emacs-lisp

     (defun reload-config () (interactive) (load-file "~/.emacs.d/init.el"))

   #+END_SRC

** Store backup / autosave files in tmp

   #+BEGIN_SRC emacs-lisp

     (setq backup-directory-alist
	   `((".*" . ,temporary-file-directory)))
     (setq auto-save-file-name-transforms
	   `((".*" ,temporary-file-directory t)))

   #+END_SRC
   
** Use PATH variable from shell

   #+BEGIN_SRC emacs-lisp

     (use-package exec-path-from-shell
       :ensure t
       :config (exec-path-from-shell-initialize))

   #+END_SRC
  
   
* UI

** Font 

   #+BEGIN_SRC emacs-lisp

     (set-frame-font "Hack Nerd Font 11" nil t)

   #+END_SRC

** Modeline 

   #+BEGIN_SRC emacs-lisp

     (use-package minions
       :ensure t
       :defer 0.1
       :config
       (setq minions-mode-line-lighter "[+]")
       (minions-mode 1))

     (use-package mood-line
       :ensure t
       :defer 0.1
       :after minions
       :config
       (defun mood-line-segment-major-mode ()
	 "Displays the current major mode in the mode-line."
	 (concat (format-mode-line minions-mode-line-modes 'mood-line-major-mode) "  "))
       (mood-line-mode)
       )

   #+END_SRC

** Org-Mode
*** Bullets

    #+BEGIN_SRC emacs-lisp

      (use-package org-bullets
	:ensure t
	:config (add-hook 'org-mode-hook (lambda() (org-bullets-mode 1))))

    #+END_SRC

** Dashboard

   #+BEGIN_SRC emacs-lisp

     (use-package dashboard
       :ensure t
       :init (progn
	       (setq dashboard-items '((recents . 1)
				       (projects . 1)))
	       (setq dashboard-show-shortcuts nil)
	       (setq dashboard-center-content nil)
	       (setq dashboard-set-file-icons t)
	       (setq dashboard-set-heading-icons t)
	       )
       :config (dashboard-setup-startup-hook)
       )

   #+END_SRC
   
   
* General Packages

** Evil 

   #+BEGIN_SRC emacs-lisp

     (use-package evil
       :ensure t
       :init
       (setq evil-want-keybinding nil)
       :config 
       (evil-ex-define-cmd "q" 'kill-this-buffer)
       (evil-ex-define-cmd "quit" 'evil-quit)
       (evil-define-key 'normal org-mode-map (kbd "C-SPC") 'org-cycle)
       (evil-mode 1))

     (use-package evil-collection
       :ensure t
       :after evil
       :config (evil-collection-init))

   #+END_SRC

** Helm

   #+BEGIN_SRC emacs-lisp

     (use-package counsel
       :ensure t
       :after ivy
       :config (counsel-mode))

     (use-package ivy
       :ensure t
       :defer 0.1
       :diminish
       :bind (("C-c C-r" . ivy-resume)
	      ("C-x B" . ivy-switch-buffer-other-window))
       :custom
       (ivy-count-format "(%d/%d) ")
       (ivy-use-virtual-buffers t)
       :config (ivy-mode))

     (use-package ivy-rich
       :ensure t
       :init (ivy-rich-mode 1))

     (use-package all-the-icons-ivy-rich
       :ensure t
       :init (all-the-icons-ivy-rich-mode 1))

     (use-package swiper
       :after ivy
       :bind (("C-s" . swiper)
	      ("C-r" . swiper)))

     (setq ivy-virtual-abbreviate 'abbreviate
	   ivy-rich-path-style 'abbrev
	   all-the-icons-ivy-rich-icon-size 1.1)

   #+END_SRC
   
** Treemacs 

   #+BEGIN_SRC emacs-lisp

     (use-package treemacs
       :ensure t
       :defer t
       :init
       (with-eval-after-load 'winum
	 (define-key winum-keymap (kbd "M-0") #'treemacs-select-window))
       :config
       (progn
	 (setq treemacs-collapse-dirs                 (if treemacs-python-executable 3 0)
	       treemacs-deferred-git-apply-delay      0.5
	       treemacs-directory-name-transformer    #'identity
	       treemacs-display-in-side-window        t
	       treemacs-eldoc-display                 t
	       treemacs-file-event-delay              5000
	       treemacs-file-extension-regex          treemacs-last-period-regex-value
	       treemacs-file-follow-delay             0.2
	       treemacs-file-name-transformer         #'identity
	       treemacs-follow-after-init             t
	       treemacs-git-command-pipe              ""
	       treemacs-goto-tag-strategy             'refetch-index
	       treemacs-indentation                   2
	       treemacs-indentation-string            " "
	       treemacs-is-never-other-window         nil
	       treemacs-max-git-entries               5000
	       treemacs-missing-project-action        'ask
	       treemacs-move-forward-on-expand        nil
	       treemacs-no-png-images                 nil
	       treemacs-no-delete-other-windows       t
	       treemacs-project-follow-cleanup        nil
	       treemacs-persist-file                  (expand-file-name ".cache/treemacs-persist" user-emacs-directory)
	       treemacs-position                      'left
	       treemacs-read-string-input             'from-child-frame
	       treemacs-recenter-distance             0.1
	       treemacs-recenter-after-file-follow    nil
	       treemacs-recenter-after-tag-follow     nil
	       treemacs-recenter-after-project-jump   'always
	       treemacs-recenter-after-project-expand 'on-distance
	       treemacs-show-cursor                   nil
	       treemacs-show-hidden-files             t
	       treemacs-silent-filewatch              nil
	       treemacs-silent-refresh                nil
	       treemacs-sorting                       'alphabetic-asc
	       treemacs-space-between-root-nodes      t
	       treemacs-tag-follow-cleanup            t
	       treemacs-tag-follow-delay              1.5
	       treemacs-user-mode-line-format         nil
	       treemacs-user-header-line-format       nil
	       treemacs-width                         35
	       treemacs-workspace-switch-cleanup      nil)

	 ;; The default width and height of the icons is 22 pixels. If you are
	 ;; using a Hi-DPI display, uncomment this to double the icon size.
	 ;;(treemacs-resize-icons 44)

	 (treemacs-follow-mode t)
	 (treemacs-filewatch-mode t)
	 (treemacs-fringe-indicator-mode 'always)
	 (pcase (cons (not (null (executable-find "git")))
		      (not (null treemacs-python-executable)))
	   (`(t . t)
	    (treemacs-git-mode 'deferred))
	   (`(t . _)
	    (treemacs-git-mode 'simple))))
       :bind
       (:map global-map
	     ("M-0"       . treemacs-select-window)
	     ("C-x t 1"   . treemacs-delete-other-windows)
	     ("C-x t t"   . treemacs)
	     ("C-x t B"   . treemacs-bookmark)
	     ("C-x t C-t" . treemacs-find-file)
	     ("C-x t M-t" . treemacs-find-tag)))

     (use-package treemacs-evil
       :after (treemacs evil)
       :ensure t)

     (use-package treemacs-projectile
       :after (treemacs projectile)
       :ensure t)

     (use-package treemacs-icons-dired
       :after (treemacs dired)
       :ensure t
       :config (treemacs-icons-dired-mode))

   #+END_SRC
   
** Which Key
   
   This package displays available keybindings in a popup window.

   #+BEGIN_SRC emacs-lisp

     (use-package which-key
       :ensure t
       :config (which-key-mode))

   #+END_SRC

** Smartparens
   
   This package enables autocompletion of brackets.

   #+BEGIN_SRC emacs-lisp

     (use-package smartparens
       :ensure t
       :config (smartparens-global-mode))

   #+END_SRC

** MagIt
   
   This package provides a Git client.

   #+BEGIN_SRC emacs-lisp

     (use-package magit
       :ensure t)

   #+END_SRC

** VTerm
   
   This package provides a terminal emulator.

   #+BEGIN_SRC emacs-lisp

     (use-package vterm 
       :ensure t)

   #+END_SRC
   
** Centaur tabs
   
   Enables tabs

   #+BEGIN_SRC emacs-lisp

     (use-package centaur-tabs
       :ensure t
       :config 
       (centaur-tabs-enable-buffer-reordering)
       (setq centaur-tabs-adjust-buffer-order 'left)
       (setq centaur-tabs-height 32)
       (setq centaur-tabs-set-icons t)
       (setq centaur-tabs-gray-out-icons 'buffer)
       (setq centaur-tabs-style "bar")
       (setq centaur-tabs-set-bar 'left)
       (setq centaur-tabs-modified-marker "•")
       (centaur-tabs-headline-match)
       (centaur-tabs-group-by-projectile-project)
       (centaur-tabs-mode t)
       (defun centaur-tabs-buffer-groups ()
	 "`centaur-tabs-buffer-groups' control buffers' group rules.

      Group centaur-tabs with mode if buffer is derived from `eshell-mode' `emacs-lisp-mode' `dired-mode' `org-mode' `magit-mode'.
      All buffer name start with * will group to \"Emacs\".
      Other buffer group by `centaur-tabs-get-group-name' with project name."
	 (list
	  (cond
	   ;; ((not (eq (file-remote-p (buffer-file-name)) nil))
	   ;; "Remote")
	   ((or (string-equal "*" (substring (buffer-name) 0 1))
		(memq major-mode '(magit-process-mode
				   magit-status-mode
				   magit-diff-mode
				   magit-log-mode
				   magit-file-mode
				   magit-blob-mode
				   magit-blame-mode
				   )))
	    "Emacs")
	   ((derived-mode-p 'prog-mode)
	    "Editing")
	   ((derived-mode-p 'dired-mode)
	    "Dired")
	   ((memq major-mode '(helpful-mode
			       help-mode))
	    "Help")
	   ((memq major-mode '(org-mode
			       org-agenda-clockreport-mode
			       org-src-mode
			       org-agenda-mode
			       org-beamer-mode
			       org-indent-mode
			       org-bullets-mode
			       org-cdlatex-mode
			       org-agenda-log-mode
			       diary-mode))
	    "OrgMode")
	   (t
	    (centaur-tabs-get-group-name (current-buffer))))))
       :hook
       (dashboard-mode . centaur-tabs-local-mode)
       )

     (define-key evil-normal-state-map (kbd "g t") 'centaur-tabs-forward)
     (define-key evil-normal-state-map (kbd "g T") 'centaur-tabs-backward)

   #+END_SRC
   
   
* Programming Packages

  #+BEGIN_SRC emacs-lisp

    (setq gc-cons-threshold (* 100 1024 1024)
	  read-process-output-max (* 1024 1024)
	  company-minimum-prefix-length 1
	  lsp-lens-enable t
	  lsp-signature-auto-activate nil)

    (global-set-key (kbd "C-x o") 'ace-window)
    (global-set-key [f10] 'recompile)

  #+END_SRC

** LSP Mode

   Set up the client for Language Server Protocol. Used to provide IDE like functionality.

   To add an lsp server, use the *M-x lsp-install-server* to install the required packages, or follow documentation.

   Documentation: https://emacs-lsp.github.io/lsp-mode/page/languages/

   LSP prefix: "C-l"

   #+BEGIN_SRC emacs-lisp

     (use-package lsp-mode
       :ensure t
       :init (setq lsp-keymap-prefix "C-l")
       :hook (lsp-mode . lsp-enable-which-key-integration)
       :commands (lsp lsp-deferred))

   #+END_SRC
    
   Some missing IDE functionality includes: fancy sideline, popup documentation, code peeking, etc.
   Including the lsp-ui package adds these to Emacs.

   #+BEGIN_SRC emacs-lisp

     (use-package lsp-ui
       :ensure t
       :after lsp
       :commands lsp-ui-mode
       :config
       (define-key lsp-ui-mode-map [remap xref-find-definitions] #'lsp-ui-peek-find-definitions)
       (define-key lsp-ui-mode-map [remap xref-find-references] #'lsp-ui-peek-find-references)
       )

   #+END_SRC

   Configuration variables for lsp-ui

   #+BEGIN_SRC emacs-lisp

     (setq lsp-ui-doc-enable t
	   lsp-ui-doc-use-childframe t
	   lsp-ui-doc-position 'top
	   lsp-ui-doc-include-signature t
	   lsp-ui-sideline-enable nil
	   lsp-ui-flycheck-enable t
	   lsp-ui-flycheck-list-position 'right
	   lsp-ui-flycheck-live-reporting t
	   lsp-ui-peek-enable t
	   lsp-ui-peek-list-width 60
	   lsp-ui-peek-peek-height 25
	   )


   #+END_SRC
    
   Since we are using Ivy, we need to enable Ivy support.

   #+BEGIN_SRC emacs-lisp

     (use-package lsp-ivy
       :ensure t
       :commands lsp-ivy-workspace-symbol
       :config (global-set-key ))

   #+END_SRC
    
   As well as enable Treemacs support.

   #+BEGIN_SRC emacs-lisp

     (use-package lsp-treemacs 
       :ensure t 
       :commands lsp-treemacs-errors-list)

   #+END_SRC

   Other packages can be added here if required (ex: dap-mode support for debugger)

** Syntax Checking

   We are going to be using FlyCheck. This package provides syntax checking with minimal configuration required.
   Another alternative available on MELPA is FlyMake.

   To see which languages are supported: https://www.flycheck.org/en/latest/languages.html#flycheck-languages

   #+BEGIN_SRC emacs-lisp

     (use-package flycheck
       :ensure t
       :init (global-flycheck-mode))

   #+END_SRC

** Auto Complete

   We are going to use company-mode as it integrates nicely with LSP mode, and doesn't require any language specific configuration.

   #+BEGIN_SRC emacs-lisp

     (use-package company
       :ensure t
       :config 
       (setq company-idle-delay 0)
       (global-company-mode 1)
       )

   #+END_SRC

** Debugger

   https://emacs-lsp.github.io/dap-mode/page/configuration/

   #+BEGIN_SRC emacs-lisp

     (use-package dap-mode
       :ensure t
       :config (setq dap-auto-configure-features '(sessions locals controls tooltip)))

     (require 'dap-cpptools)

   #+END_SRC

** Project Management

   Project integration library for Emacs.

   #+BEGIN_SRC emacs-lisp

     (use-package projectile
       :ensure t
       :config (projectile-mode +1)
       (define-key projectile-mode-map (kbd "C-c p") 'projectile-command-map))

   #+END_SRC

** Code Snippets

   #+BEGIN_SRC emacs-lisp

     (use-package yasnippet
       :ensure t
       :config (setq yas-snippets-dirs '("~/.emacs.d/snippets"))
       (yas-global-mode 1))

     (global-set-key (kbd "C-c i") 'yas-insert-snippet)

   #+END_SRC

** Language Specific Configuration

*** CMake

    #+BEGIN_SRC emacs-lisp

      (use-package cmake-mode
	:ensure t
	:mode ("CMakeLists\\.txt\\'" "\\.cmake\\'"))

      (use-package cmake-font-lock
	:ensure t
	:after cmake-mode
	:config (cmake-font-lock-activate))


	(add-hook 'cmake-mode 'lsp-deferred)

    #+END_SRC
 
*** C/C++

    #+BEGIN_SRC emacs-lisp

      (add-hook 'c-mode-hook 'lsp-deferred)
      (add-hook 'c++-mode-hook 'lsp-deferred)

    #+END_SRC
 
*** Lua

    #+BEGIN_SRC emacs-lisp

      (add-hook 'lua-mode-hook 'lsp-deferred)

    #+END_SRC
 
*** PHP

    #+BEGIN_SRC emacs-lisp

      (use-package php-mode
	:ensure t
	:mode ("\\.php\\'" . php-mode))

      (use-package phpunit
	:ensure t)

      (add-hook 'php-mode-hook 'lsp-deferred)

    #+END_SRC
 
    
*** Rust

    #+BEGIN_SRC emacs-lisp

      (use-package rust-mode 
	:ensure t)

      (add-hook 'rust-mode-hook 'lsp-deferred)

      #+END_SRC
 
    
    


config.el
(setq package-archives
      '(("elpy" . "http://jorgenschaefer.github.io/packages/")
	("melpa" . "https://melpa.org/packages/")
	("gnu" . "http://elpa.gnu.org/packages/")
	("melpa-stable" . "https://stable.melpa.org/packages/")))

(setq default-tab-width 2)

(setq inhibit-startup-screen t)

(tool-bar-mode -1)

(menu-bar-mode -1)

(toggle-scroll-bar -1)

(setq visible-bell nil)

(show-paren-mode 1)

(global-hl-line-mode +1)

(add-hook 'prog-mode-hook 'display-line-numbers-mode)

(defun reload-config () (interactive) (load-file "~/.emacs.d/init.el"))

(setq backup-directory-alist
      `((".*" . ,temporary-file-directory)))
(setq auto-save-file-name-transforms
      `((".*" ,temporary-file-directory t)))

(use-package exec-path-from-shell
  :ensure t
  :config (exec-path-from-shell-initialize))

(set-frame-font "Hack Nerd Font 11" nil t)

(use-package minions
  :ensure t
  :defer 0.1
  :config
  (setq minions-mode-line-lighter "[+]")
  (minions-mode 1))

(use-package mood-line
  :ensure t
  :defer 0.1
  :after minions
  :config
  (defun mood-line-segment-major-mode ()
    "Displays the current major mode in the mode-line."
    (concat (format-mode-line minions-mode-line-modes 'mood-line-major-mode) "  "))
  (mood-line-mode)
  )

(use-package org-bullets
  :ensure t
  :config (add-hook 'org-mode-hook (lambda() (org-bullets-mode 1))))

(use-package dashboard
  :ensure t
  :init (progn
	  (setq dashboard-items '((recents . 1)
				  (projects . 1)))
	  (setq dashboard-show-shortcuts nil)
	  (setq dashboard-center-content nil)
	  (setq dashboard-set-file-icons t)
	  (setq dashboard-set-heading-icons t)
	  )
  :config (dashboard-setup-startup-hook)
  )

(use-package evil
  :ensure t
  :init
  (setq evil-want-keybinding nil)
  :config 
  (evil-ex-define-cmd "q" 'kill-this-buffer)
  (evil-ex-define-cmd "quit" 'evil-quit)
  (evil-define-key 'normal org-mode-map (kbd "C-SPC") 'org-cycle)
  (evil-mode 1))

(use-package evil-collection
  :ensure t
  :after evil
  :config (evil-collection-init))

(use-package counsel
  :ensure t
  :after ivy
  :config (counsel-mode))

(use-package ivy
  :ensure t
  :defer 0.1
  :diminish
  :bind (("C-c C-r" . ivy-resume)
	 ("C-x B" . ivy-switch-buffer-other-window))
  :custom
  (ivy-count-format "(%d/%d) ")
  (ivy-use-virtual-buffers t)
  :config (ivy-mode))

(use-package ivy-rich
  :ensure t
  :init (ivy-rich-mode 1))

(use-package all-the-icons-ivy-rich
  :ensure t
  :init (all-the-icons-ivy-rich-mode 1))

(use-package swiper
  :after ivy
  :bind (("C-s" . swiper)
	 ("C-r" . swiper)))

(setq ivy-virtual-abbreviate 'abbreviate
      ivy-rich-path-style 'abbrev
      all-the-icons-ivy-rich-icon-size 1.1)

(use-package treemacs
  :ensure t
  :defer t
  :init
  (with-eval-after-load 'winum
    (define-key winum-keymap (kbd "M-0") #'treemacs-select-window))
  :config
  (progn
    (setq treemacs-collapse-dirs                 (if treemacs-python-executable 3 0)
	  treemacs-deferred-git-apply-delay      0.5
	  treemacs-directory-name-transformer    #'identity
	  treemacs-display-in-side-window        t
	  treemacs-eldoc-display                 t
	  treemacs-file-event-delay              5000
	  treemacs-file-extension-regex          treemacs-last-period-regex-value
	  treemacs-file-follow-delay             0.2
	  treemacs-file-name-transformer         #'identity
	  treemacs-follow-after-init             t
	  treemacs-git-command-pipe              ""
	  treemacs-goto-tag-strategy             'refetch-index
	  treemacs-indentation                   2
	  treemacs-indentation-string            " "
	  treemacs-is-never-other-window         nil
	  treemacs-max-git-entries               5000
	  treemacs-missing-project-action        'ask
	  treemacs-move-forward-on-expand        nil
	  treemacs-no-png-images                 nil
	  treemacs-no-delete-other-windows       t
	  treemacs-project-follow-cleanup        nil
	  treemacs-persist-file                  (expand-file-name ".cache/treemacs-persist" user-emacs-directory)
	  treemacs-position                      'left
	  treemacs-read-string-input             'from-child-frame
	  treemacs-recenter-distance             0.1
	  treemacs-recenter-after-file-follow    nil
	  treemacs-recenter-after-tag-follow     nil
	  treemacs-recenter-after-project-jump   'always
	  treemacs-recenter-after-project-expand 'on-distance
	  treemacs-show-cursor                   nil
	  treemacs-show-hidden-files             t
	  treemacs-silent-filewatch              nil
	  treemacs-silent-refresh                nil
	  treemacs-sorting                       'alphabetic-asc
	  treemacs-space-between-root-nodes      t
	  treemacs-tag-follow-cleanup            t
	  treemacs-tag-follow-delay              1.5
	  treemacs-user-mode-line-format         nil
	  treemacs-user-header-line-format       nil
	  treemacs-width                         35
	  treemacs-workspace-switch-cleanup      nil)

    ;; The default width and height of the icons is 22 pixels. If you are
    ;; using a Hi-DPI display, uncomment this to double the icon size.
    ;;(treemacs-resize-icons 44)

    (treemacs-follow-mode t)
    (treemacs-filewatch-mode t)
    (treemacs-fringe-indicator-mode 'always)
    (pcase (cons (not (null (executable-find "git")))
		 (not (null treemacs-python-executable)))
      (`(t . t)
       (treemacs-git-mode 'deferred))
      (`(t . _)
       (treemacs-git-mode 'simple))))
  :bind
  (:map global-map
	("M-0"       . treemacs-select-window)
	("C-x t 1"   . treemacs-delete-other-windows)
	("C-x t t"   . treemacs)
	("C-x t B"   . treemacs-bookmark)
	("C-x t C-t" . treemacs-find-file)
	("C-x t M-t" . treemacs-find-tag)))

(use-package treemacs-evil
  :after (treemacs evil)
  :ensure t)

(use-package treemacs-projectile
  :after (treemacs projectile)
  :ensure t)

(use-package treemacs-icons-dired
  :after (treemacs dired)
  :ensure t
  :config (treemacs-icons-dired-mode))

(use-package which-key
  :ensure t
  :config (which-key-mode))

(use-package smartparens
  :ensure t
  :config (smartparens-global-mode))

(use-package magit
  :ensure t)

(use-package vterm 
  :ensure t)

(use-package centaur-tabs
  :ensure t
  :config 
  (centaur-tabs-enable-buffer-reordering)
  (setq centaur-tabs-adjust-buffer-order 'left)
  (setq centaur-tabs-height 32)
  (setq centaur-tabs-set-icons t)
  (setq centaur-tabs-gray-out-icons 'buffer)
  (setq centaur-tabs-style "bar")
  (setq centaur-tabs-set-bar 'left)
  (setq centaur-tabs-modified-marker "•")
  (centaur-tabs-headline-match)
  (centaur-tabs-group-by-projectile-project)
  (centaur-tabs-mode t)
  (defun centaur-tabs-buffer-groups ()
    "`centaur-tabs-buffer-groups' control buffers' group rules.

 Group centaur-tabs with mode if buffer is derived from `eshell-mode' `emacs-lisp-mode' `dired-mode' `org-mode' `magit-mode'.
 All buffer name start with * will group to \"Emacs\".
 Other buffer group by `centaur-tabs-get-group-name' with project name."
    (list
     (cond
      ;; ((not (eq (file-remote-p (buffer-file-name)) nil))
      ;; "Remote")
      ((or (string-equal "*" (substring (buffer-name) 0 1))
	   (memq major-mode '(magit-process-mode
			      magit-status-mode
			      magit-diff-mode
			      magit-log-mode
			      magit-file-mode
			      magit-blob-mode
			      magit-blame-mode
			      )))
       "Emacs")
      ((derived-mode-p 'prog-mode)
       "Editing")
      ((derived-mode-p 'dired-mode)
       "Dired")
      ((memq major-mode '(helpful-mode
			  help-mode))
       "Help")
      ((memq major-mode '(org-mode
			  org-agenda-clockreport-mode
			  org-src-mode
			  org-agenda-mode
			  org-beamer-mode
			  org-indent-mode
			  org-bullets-mode
			  org-cdlatex-mode
			  org-agenda-log-mode
			  diary-mode))
       "OrgMode")
      (t
       (centaur-tabs-get-group-name (current-buffer))))))
  :hook
  (dashboard-mode . centaur-tabs-local-mode)
  )

(define-key evil-normal-state-map (kbd "g t") 'centaur-tabs-forward)
(define-key evil-normal-state-map (kbd "g T") 'centaur-tabs-backward)

(setq gc-cons-threshold (* 100 1024 1024)
      read-process-output-max (* 1024 1024)
      company-minimum-prefix-length 1
      lsp-lens-enable t
      lsp-signature-auto-activate nil)

(global-set-key (kbd "C-x o") 'ace-window)
(global-set-key [f10] 'recompile)

(use-package lsp-mode
  :ensure t
  :init (setq lsp-keymap-prefix "C-l")
  :hook (lsp-mode . lsp-enable-which-key-integration)
  :commands (lsp lsp-deferred))

(use-package lsp-ui
  :ensure t
  :after lsp
  :commands lsp-ui-mode
  :config
  (define-key lsp-ui-mode-map [remap xref-find-definitions] #'lsp-ui-peek-find-definitions)
  (define-key lsp-ui-mode-map [remap xref-find-references] #'lsp-ui-peek-find-references)
  )

(setq lsp-ui-doc-enable t
      lsp-ui-doc-use-childframe t
      lsp-ui-doc-position 'top
      lsp-ui-doc-include-signature t
      lsp-ui-sideline-enable nil
      lsp-ui-flycheck-enable t
      lsp-ui-flycheck-list-position 'right
      lsp-ui-flycheck-live-reporting t
      lsp-ui-peek-enable t
      lsp-ui-peek-list-width 60
      lsp-ui-peek-peek-height 25
      )

(use-package lsp-ivy
  :ensure t
  :commands lsp-ivy-workspace-symbol
  :config (global-set-key ))

(use-package lsp-treemacs 
  :ensure t 
  :commands lsp-treemacs-errors-list)

(use-package flycheck
  :ensure t
  :init (global-flycheck-mode))

(use-package company
  :ensure t
  :config 
  (setq company-idle-delay 0)
  (global-company-mode 1)
  )

(use-package dap-mode
  :ensure t
  :config (setq dap-auto-configure-features '(sessions locals controls tooltip)))

(require 'dap-cpptools)

(use-package projectile
  :ensure t
  :config (projectile-mode +1)
  (define-key projectile-mode-map (kbd "C-c p") 'projectile-command-map))

(use-package yasnippet
  :ensure t
  :config (setq yas-snippets-dirs '("~/.emacs.d/snippets"))
  (yas-global-mode 1))

(global-set-key (kbd "C-c i") 'yas-insert-snippet)

(use-package cmake-mode
  :ensure t
  :mode ("CMakeLists\\.txt\\'" "\\.cmake\\'"))

(use-package cmake-font-lock
  :ensure t
  :after cmake-mode
  :config (cmake-font-lock-activate))


  (add-hook 'cmake-mode 'lsp-deferred)

(add-hook 'c-mode-hook 'lsp-deferred)
(add-hook 'c++-mode-hook 'lsp-deferred)

(add-hook 'lua-mode-hook 'lsp-deferred)

(use-package php-mode
  :ensure t
  :mode ("\\.php\\'" . php-mode))

(use-package phpunit
  :ensure t)

(add-hook 'php-mode-hook 'lsp-deferred)

(use-package rust-mode 
  :ensure t)

(add-hook 'rust-mode-hook 'lsp-deferred)
